### ISSUE TITLE
STJ: JsonNode.ReplaceWith used with an array will result in a node that identifies as array but is not a JsonArray node

### ISSUE THREAD

**Issue by elgonzo** [dotnet/runtime#100470]
### Description

Using `JsonNode.ReplaceWith` with an array will result in a JsonNode instance that identifies as an array through its `JsonNode.GetValueKind` method, but is actually no JsonArray node.

### Reproduction Steps

var rootNode = new JsonObject();
var childNode = new JsonObject();
rootNode["child"] = childNode;

childNode.ReplaceWith(Array.Empty<int>());

Console.WriteLine(rootNode["child"].GetValueKind());
Console.WriteLine(rootNode["child"].AsArray());

### Expected behavior

Either `rootNode["child"].GetValueKind()` returning `JsonValueKind.Array` and `rootNode["child"].AsArray()` returning a JsonArray instance.
Or alternatively, `rootNode["child"].GetValueKind()` _not_ returning `JsonValueKind.Array` and `rootNode["child"].AsArray()`.

### Actual behavior

`rootNode["child"].GetValueKind()` returns `JsonValueKind.Array`, yet `rootNode["child"].AsArray()` throws.

### Known Workarounds

Don't use `JsonNode.AsArray` method, or preferably don't pass anything to `JsonNode.ReplaceWith` that's not a JsonNode instance nor a primitive type nor string (despite the neither the API nor the documentation indicating that the ReplaceWith method has any such constraints).

### Configuration

.NET 8.0.2
System.Text.Json 9.0.0-preview.2.24128.5

**Comment by eiriktsarpalis** [comment]
I don't believe this is an issue with the `ReplaceWith` method per se, but an inherent quirk of `JsonNode` design which is capable of encapsulating arbitrary .NET objects. For example:

```C#
JsonNode node = JsonValue.Create(Array.Empty<int>());
node.AsArray(); // Same error as above
```

I don't believe that this is something we could fix.

**Comment by elgonzo** [comment]
Using the `ReplaceWith` API, how would i replace a child node with a json array for example?

I assumed previously that i could pass a JsonArray instance to `ReplaceWith`, but as i experienced in the meantime, `ReplaceWith` rejects any JsonNode instance passed to it. It seems odd to me that the `ReplaceWith` API doesn't support scenarios where the replacement node is not supposed to be a json value, and i can't tell whether this would be a mistake in the API design or whether this a deliberate restriction for some good reason.

**Comment by eiriktsarpalis** [comment]
> but as i experienced in the meantime, `ReplaceWith` rejects any JsonNode instance passed to it.

Can you give an example? This works for me

```C#
var rootNode = new JsonObject();
var childNode = new JsonObject();
rootNode["child"] = childNode;

childNode.ReplaceWith(new JsonArray { 1, 2, 3 });

Console.WriteLine($"Value kind: {rootNode["child"].GetValueKind()}");
Console.WriteLine(rootNode["child"].GetType());
Console.WriteLine(rootNode["child"].AsArray());
```

**Comment by elgonzo** [comment]
My apologies, you are right. I can pass JsonArray to ReplaceWith without problems.

I wonder why i saw an exception when i tried this yesterday on my PC. I am now on my laptop (without access to the PC) and needed to recreate the test project and everything works fine. Looking at the commit history of the JsonNode.cs file, i noticed there was once a time where ReplaceWith was doing `JsonValue.Create(value)`, something that could explain the exception i saw and which you already corrected in November 2023. I suspect i somehow half-asleep selected a STJ 8 preview version instead of the STJ 9 preview during my trials yesterday, but that's not something you need to be concerned about - that's entirely on me. Again, my apologies for creating noise :-(

**Comment by eiriktsarpalis** [comment]
No worries at all. Let us know in case you're able to reproduce unexpected behavior and we can reopen this.
